CDAP-8014 Cache RouteDestination as an optimization

Denton-L opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-8014
Bamboo: http://builds.cask.co/browse/CDAP-DUT5249

Denton-L committed:
CDAP-8014 Cache RouteDestination as an optimization

This resolves the TODO that was in the code about optimization.

sreevatsanraman commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java:
```diff
@@ -41,6 +46,9 @@
   public static final RouteDestination STREAMS_SERVICE = new RouteDestination(Constants.Service.STREAMS);
   public static final RouteDestination PREVIEW_HTTP = new RouteDestination(Constants.Service.PREVIEW_HTTP);
 
+  private Map<MultiKey, RouteDestination> _routeDestinationCache = Collections.synchronizedMap(
```
remove the underscore prefix in the variable name.

Denton-L replied:
Should we remove item 15 from the coding standards page on Confluence?

Denton-L committed:
Remove underscore from variable name

Denton-L commented:
Should we remove item 15 from the coding standards page on Confluence?

bdmogal commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java:
```diff
@@ -41,6 +46,9 @@
   public static final RouteDestination STREAMS_SERVICE = new RouteDestination(Constants.Service.STREAMS);
   public static final RouteDestination PREVIEW_HTTP = new RouteDestination(Constants.Service.PREVIEW_HTTP);
 
+  private Map<MultiKey, RouteDestination> routeDestinationCache = Collections.synchronizedMap(
```
final

bdmogal commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java:
```diff
@@ -83,8 +91,12 @@ private RouteDestination getV3RoutingService(String [] uriParts, AllowedMethod r
-      return new RouteDestination(serviceName, version);
+      MultiKey cacheKey = new MultiKey(serviceName, version);
+
+      if (!routeDestinationCache.containsKey(cacheKey)) {
```
hmmm, why do you need a synchronized map? Also, if you do a ``containsKey`` + ``put``, its not really synchronized. Use ``putIfAbsent``.

Denton-L committed:
Make routeDestinationCache final

Denton-L committed:
Use ConcurrentMap instead

Denton-L committed:
Fix incorrect logic

Before this, using putIfAbsent would cause an object to be created
anyway, rendering this feature useless.

Denton-L commented:
@bdmogal I switched to ConcurrentMap but, unfortunately, I cannot use `putIfAbsent` because this would cause an object to be created anyway which is exactly what we are trying to prevent. 

EDIT: Realistically, I don't think that there would be any synchronization problems. Even if the "wrong" object were placed in the map, it would still work as expected, I believe.

chtyim commented on cdap-gateway/src/main/java/co/cask/cdap/gateway/router/RouterPathLookup.java:
```diff
@@ -83,8 +89,12 @@ private RouteDestination getV3RoutingService(String [] uriParts, AllowedMethod r
       String version = uriParts[6];
-      // TODO: Optimize this instead of creating new object every time
-      return new RouteDestination(serviceName, version);
+      MultiKey cacheKey = new MultiKey(serviceName, version);
```
This is not an optimization at all. It used to creating one `RouteDestination` object. Now it is creation of one `MultiKey`, plus two lookup in a concurrent map (`containsKey` and `get`). I don't think we need to change anything here. Simply remove the TODO and mark the CDAP-8014 as won't fix.

Shouldn't do any optimization without actual perf test backing it.
